### ISSUE TITLE
llvm-as doesn't accept `-defsym something`

### ISSUE THREAD

**Issue by jeremyd2019** [msys2/MINGW-packages#9179]
When building mingw-w64-perl-win32-api:
```
  as  -defsym HAS_TIMES=1 -defsym HAVE_INTERP_INTERN=1 -defsym MULTIPLICITY=1 -defsym PERLIO_LAYERS=1 -defsym PERL_IMPLICIT_CONTEXT=1 -defsym PERL_IMPLICIT_SYS=1 -defsym USE_64_BIT_INT=1 -defsym USE_ITHREADS=1 -defsym USE_LARGE_FILES=1 -defsym USE_LOCALE_COLLATE=1 -defsym USE_LOCALE_NUMERIC=1 -defsym USE_LOCALE_TIME=1 -defsym USE_PERLIO=1  call_asm_x64_gnu.s -o call_asm_x64_gnu.o
  as: error: no such file or directory: 'HAS_TIMES=1'
  as: error: no such file or directory: 'HAVE_INTERP_INTERN=1'
  as: error: no such file or directory: 'MULTIPLICITY=1'
  as: error: no such file or directory: 'PERLIO_LAYERS=1'
  as: error: no such file or directory: 'PERL_IMPLICIT_CONTEXT=1'
  as: error: no such file or directory: 'PERL_IMPLICIT_SYS=1'
  as: error: no such file or directory: 'USE_64_BIT_INT=1'
  as: error: no such file or directory: 'USE_ITHREADS=1'
  as: error: no such file or directory: 'USE_LARGE_FILES=1'
  as: error: no such file or directory: 'USE_LOCALE_COLLATE=1'
  as: error: no such file or directory: 'USE_LOCALE_NUMERIC=1'
  as: error: no such file or directory: 'USE_LOCALE_TIME=1'
  as: error: no such file or directory: 'USE_PERLIO=1'
  mingw32-make: *** [Makefile:1177: call_asm_x64_gnu.o] Error 1
```

Is this not supported with llvm?  I've generally always seen `-x assembler-with-cpp` for things like this

**Comment by jeremyd2019** [comment]
/cc @mstorsjo 

It so happens that the .s file in question doesn't even use any of the defines (the 32-bit one uses one of them)

**Comment by jeremyd2019** [comment]
I guess `as` isn't `llvm-as` but actually an alias of `clang`.  In any event, it isn't taking that option that GNU as does.

**Comment by mstorsjo** [comment]
Yep, `llvm-as` is an entirely different thing. Providing an `as` compatible tool is a bit of a mess; there's `llvm-mc` which does exactly the same thing, assembly to object file, but it's not meant to be used as an end user tool, it's primarily used by the llvm tests.

And then clang mostly implements an interface which is close to `as`, but not exactly, e.g. with `as` you can do `as foo.s`  (without `-c`) and it'd produce an object file, while `clang` would try to assemble+link. With GCC and `as`, there's a clear distinction and you can pass options to it with `-Wa`, or pass those options directly to `as`. With Clang, there's no actual interface to the intermediate thing, so Clang itself parses some of the `-Wa` options - e.g. https://github.com/llvm/llvm-project/blob/635288d215b1147e624cf37f9c5f66ae35c15b74/clang/lib/Driver/ToolChains/Clang.cpp#L2553-L2573.

So Clang does support `-defsym`, but only in the form `-Wa,-defsym,foo=123`. And passing that to `as` is wrong. But it would work across both GCC/`as` and Clang if you'd make it call `$CC` and wrap the options in `-Wa` instead.

Making Clang recognize when it's being invoked in the form of `as` (just like it differentiates between being called `clang`, `clang++` or `clang-cl`) and then accepting these options only, without the `-Wa` prefix, would be a very welcome feature (according to me at least), but I think it'd require at least a moderate amount of changes.

**Comment by jeremyd2019** [comment]
> So Clang does support `-defsym`, but only in the form `-Wa,-defsym,foo=123`. And passing that to `as` is wrong. But it would work across both GCC/`as` and Clang if you'd make it call `$CC` and wrap the options in `-Wa` instead.

Ah, ok, I could probably make that work.  In all, this package is pretty crufty.  It looks like a half-baked attempt at making something like libffi for calling Windows APIs from perl.  Probably predates libffi though.

**Comment by mstorsjo** [comment]
Btw, I'm curious, what's the overall progress in number of packages compiling successfully (with patches) for ucrt64 and clang64 out of the total, so far?

**Comment by jeremyd2019** [comment]
@lazka could probably provide better numbers, but in mingw64.db I see 1791 unique base packages, in ucrt64.db there are 1731, and in clang64.db 1377.

**Comment by mstorsjo** [comment]
> @lazka could probably provide better numbers, but in mingw64.db I see 1791 unique base packages, in ucrt64.db there are 1731, and in clang64.db 1377.

Nice! And the rest are ones that fail for reasons that aren’t yet known - or is it just that they haven’t been tried yet?

**Comment by Biswa96** [comment]
The list is also here https://packages.msys2.org/repos

**Comment by jeremyd2019** [comment]
> The list is also here https://packages.msys2.org/repos

Cool.  I just happened to already have a script that built a set of base packages directly from a db file, so used that to output a count, but that's easier (but counts packages not base packages, so will be higher due to split packages).

**Comment by mati865** [comment]
Our `as` is clang indeed. At least one of packages using autotools was looking for it and could only find `/usr/bin/as.exe` which is obvioisly incompatible.
Having more packages ported to sane build systems we can hopefully get rid of this alias.

**Comment by jeremyd2019** [comment]
https://github.com/msys2/MINGW-packages/blob/f94d115ea2b8c249f00784158aa6f9c2fdc7f552/mingw-w64-perl-win32-api/0001-clang-as-defines.patch

**Comment by mstorsjo** [comment]
> https://github.com/msys2/MINGW-packages/blob/f94d115ea2b8c249f00784158aa6f9c2fdc7f552/mingw-w64-perl-win32-api/0001-clang-as-defines.patch

I guess that works. Would it be safer or more precise to inspect $(AS) instead of $(CC) though?

**Comment by mstorsjo** [comment]
> > https://github.com/msys2/MINGW-packages/blob/f94d115ea2b8c249f00784158aa6f9c2fdc7f552/mingw-w64-perl-win32-api/0001-clang-as-defines.patch
> 
> I guess that works. Would it be safer or more precise to inspect $(AS) instead of $(CC) though?

Oh nevermind, that’s not necessarily easily available in that form while executing the perl script I realize now.
